client/pov-recovery/src/lib.rs

```diff
@@ -105,26 +132,84 @@ pub struct RecoveryDelay {
 	pub max: Duration,
 }
 
+impl RecoveryDelay {
+	/// Produce a randomized duration between `min` and `max`.
+	fn duration(&self) -> Duration {
+		self.min + self.max.saturating_sub(self.min).mul_f64(thread_rng().gen())
+	}
+}
+
 /// Represents an outstanding block candidate.
 struct Candidate<Block: BlockT> {
 	receipt: CandidateReceipt,
 	session_index: SessionIndex,
 	block_number: NumberFor<Block>,
 	parent_hash: Block::Hash,
 	// Lazy recovery has been submitted.
+	// Should be true iff a block is either queued to be recovered or
+	// recovery is currently in progress.
 	waiting_recovery: bool,
 }
 
+struct RecoveryQueue<Block: BlockT> {
+	// Queue that keeps the hashes of blocks to be recovered.
+	recovery_queue: VecDeque<Block::Hash>,
+	// Futures that resolve when a new recovery should be started.
+	signaling_queue: FuturesUnordered<Pin<Box<dyn Future<Output = ()> + Send>>>,
+}
+
+/// Queue that is used to decide when to start PoV-recovery operations.
+impl<Block: BlockT> RecoveryQueue<Block> {
+	pub fn new() -> Self {
+		return Self { recovery_queue: VecDeque::new(), signaling_queue: FuturesUnordered::new() }
+	}
+
+	/// Add hash of a block that should go to the end of the recovery queue.
+	/// A new recovery will be signaled after `delay` has passed.
+	pub fn push_recovery(&mut self, hash: Block::Hash, delay: Duration) {
+		tracing::debug!(
+			target: LOG_TARGET,
+			block_hash = ?hash,
+			"Adding block to queue and adding new recovery slot in {:?} sec",
+			delay.as_secs(),
+		);
+		self.recovery_queue.push_back(hash);
+		self.signaling_queue.push(
+			async move {
+				Delay::new(delay).await;
+			}
+			.boxed(),
+		);
+	}
+
+	/// Get the next hash for block recovery.
+	pub async fn next_recovery(&mut self) -> Block::Hash {
+		loop {
+			if let Some(_) = self.signaling_queue.next().await {
+				if let Some(hash) = self.recovery_queue.pop_front() {
+					return hash
+				} else {
+					tracing::warn!(
+						target: LOG_TARGET,
+						"Recovery was signaled, but no candidate hash available. This is a bug."
+					);
+				};
+			}
+			futures::pending!()
+		}
+	}
```

Comment:
Hmm... can't this in theory mix up the delays? e.g.:

```rust
queue.push_recovery(BLOCK_A, Duration::from_secs(10));
queue.push_recovery(BLOCK_B, Duration::from_secs(5));
```

The 5s duration from the BLOCK_B would fire first (since this uses `FuturesUnordered`), but would pop from the front of the `recovery_queue`, so will pop BLOCK_A.

Also, perhaps it'd be better to pass the `retry_delay: Duration` and `recovery_delay: RecoveryDelay` in `new()` (since those are effectively constant after they're calculated, if I'm looking at it right?) and then just sample the `recovery_delay` internally here in the queue when it's needed?

Reply:
> Hmm... can't this in theory mix up the delays? e.g.:

Yes this is correct. Basically the delay adds only a new "signaling slot" to the signaling queue at which the next block of the queue is recovered. Insertion order should be respected, but it does not matter by which signal triggers the recovery. 

The other suggestion is nice, I think we can even get rid of retry_delay and just use the same delay one for retries and simple recoveries.

Reply:
Delays have been merged into the queue :+1: